kornia/metrics/ssim.py

```diff
@@ -75,15 +75,16 @@ def ssim(
         raise ValueError(f"img1 and img2 shapes must be the same. Got: {img1.shape} and {img2.shape}")
 
     # prepare kernel
-    kernel: torch.Tensor = get_gaussian_kernel2d((window_size, window_size), (1.5, 1.5))
+    kernel_x: torch.Tensor = get_gaussian_kernel1d(window_size, 1.5)
+    kernel_y: torch.Tensor = get_gaussian_kernel1d(window_size, 1.5)
 
     # compute coefficients
     C1: float = (0.01 * max_val) ** 2
     C2: float = (0.03 * max_val) ** 2
 
     # compute local mean per channel
-    mu1: torch.Tensor = filter2d(img1, kernel)
-    mu2: torch.Tensor = filter2d(img2, kernel)
+    mu1: torch.Tensor = filter2d_separable(img1, kernel_x, kernel_y)
+    mu2: torch.Tensor = filter2d_separable(img2, kernel_x, kernel_y)
 
     cropping_shape: List[int] = []
     if padding == 'valid':
```

Comment:
needs to fix the h and w at this branch

Reply:
btw, it doesn't have tests covering it too

Reply:
hey, sorry I wasn't familiar with the practices. I tested it myself both to make sure that it produces the same result and see the speed difference. My test code is [here](https://gist.github.com/dodobyte/9a87bd68561df4548eebfb02ac99f702).

the benchmark for different kernels is the following, image resolutions are 1024x1024, time is averaged over 1000 runs.

| Kernel  | SSIM | SSIM-Sep. | Time | Time-Sep. |
| --- | --- | --- | --- | --- |
| 11  | 0.8567 | 0.8567  | 1.55 ms|1.22 ms|
| 21  | 0.8566 | 0.8566  | 3.62 ms|1.54 ms|
| 31  | 0.8566 | 0.8566  | 7.18 ms|1.86 ms|
| 51  | 0.8566 | 0.8566  | 17.98 ms|2.49 ms|
| 101  | 0.8566 | 0.8566  | 66.95 ms|4.06 ms|

Reply:
by the way, I also adapted tensorflow's implementation and it's faster as it has one less (4) convolution calls.
I'll leave that code here as a reference as well as the benchmark.

| Kernel | Time | Time-Sep. | Time-TF |
| --- | --- | --- | --- |
| 11  | 1.55 ms|1.22 ms|0.91 ms|
| 21  | 3.62 ms|1.54 ms|1.16 ms|
| 31  | 7.18 ms|1.86 ms|1.41 ms|
| 51  | 17.98 ms|2.49 ms|1.92 ms|
| 101 | 66.95 ms|4.06 ms|3.19 ms|

```python
def ssim_tf(x, y, window=11, sigma=1.5, max_val=1.0):
	# prepare kernel
	kernel_x = get_gaussian_kernel1d(window, sigma).cuda()
	kernel_y = get_gaussian_kernel1d(window, sigma).cuda()

	c1 = (0.01 * max_val)**2
	c2 = (0.03 * max_val)**2

	# SSIM luminance measure is
	# (2 * mu_x * mu_y + c1) / (mu_x ** 2 + mu_y ** 2 + c1).
	mean0 = filter2d_separable(x, kernel_x, kernel_y)
	mean1 = filter2d_separable(y, kernel_x, kernel_y)
	num0 = mean0 * mean1 * 2.0
	den0 = mean0 ** 2 + mean1 ** 2
	luminance = (num0 + c1) / (den0 + c1)

	# SSIM contrast-structure measure is
	#   (2 * cov_{xy} + c2) / (cov_{xx} + cov_{yy} + c2).
	num1 = filter2d_separable(x * y, kernel_x, kernel_y) * 2.0
	den1 = filter2d_separable(x ** 2 + y ** 2, kernel_x, kernel_y)
	cs = (num1 - num0 + c2) / (den1 - den0 + c2)

	# SSIM score is the product of the luminance and contrast-structure measures.
	return luminance * cs
```

Reply:
Great ! If tests are passing I’d move to this implementation

Reply:
@dodobyte can you fix line 90 and add a test for it?

Reply:
sure will do that now.

Reply:
Oh, we don't have tests at all for this (related to #2231) 

But it should be like https://github.com/kornia/kornia/blob/master/test/test_metrics.py#L225

Reply:
This can be done into a subsequent PR 😄